pkg/controller/infrastructureconfig/sync_specstatus.go

```diff
@@ -0,0 +1,195 @@
+		specApiVips = &updatedInfraConfig.Spec.PlatformSpec.OpenStack.APIServerInternalIPs
+
+		statusIngressVips = &updatedInfraConfig.Status.PlatformStatus.OpenStack.IngressIPs
+		specIngressVips = &updatedInfraConfig.Spec.PlatformSpec.OpenStack.IngressIPs
+
+		statusMachineNetworks = &updatedInfraConfig.Status.PlatformStatus.OpenStack.MachineNetworks
+		specMachineNetworks = &updatedInfraConfig.Spec.PlatformSpec.OpenStack.MachineNetworks
+
+		if updatedInfraConfig.Status.PlatformStatus.OpenStack.LoadBalancer != nil &&
+			updatedInfraConfig.Status.PlatformStatus.OpenStack.LoadBalancer.Type == configv1.LoadBalancerTypeUserManaged {
+			elb = true
+		}
+
+	default:
+		// nothing to do for this platform type
+		return updatedInfraConfig, nil
+	}
+
+	if err := syncMachineNetworks(specMachineNetworks, statusMachineNetworks); err != nil {
+		return nil, fmt.Errorf("Error on syncing machine networks: %v", err)
+	}
+	if err := validateVipsWithVips(*specApiVips, *specIngressVips, elb); err != nil {
+		return nil, fmt.Errorf("Error on validating VIPs: %v", err)
+	}
+	if err := validateVipsWithMachineNetworks(*specApiVips, *specMachineNetworks); err != nil {
+		return nil, fmt.Errorf("Error on validating API VIPs and Machine Networks: %v", err)
+	}
+	if err := validateVipsWithMachineNetworks(*specIngressVips, *specMachineNetworks); err != nil {
+		return nil, fmt.Errorf("Error on validating Ingress VIPs and Machine Networks: %v", err)
+	}
+
+	if err := syncVips(specApiVips, statusApiVips); err != nil {
+		return nil, fmt.Errorf("Error on syncing API VIPs: %v", err)
+	}
+	if err := syncVips(specIngressVips, statusIngressVips); err != nil {
+		return nil, fmt.Errorf("Error on syncing Ingress VIPs: %v", err)
+	}
+
+	return updatedInfraConfig, nil
+}
+
+// syncMachineNetworks propagates machine networks from Spec to Status assuming they pass required validations.
+// In order to be valid, machine networks need to fulfil the following spec:
+//
+//   - the first machine network can never be modified
+//     -- the only exception is when Spec is empty, then adding is allowed
+//   - the number of machine networks is not limited to 2 as opposed to e.g. service networks
+//   - removal of machine networks is allowed as long as first entry is not modified
+//
+// It is possible that Status and/or Spec are empty if this is the very first time we modify this field
+// after upgrading from the version of the API that did not contain the field.
+func syncMachineNetworks(spec, status *[]configv1.CIDR) error {
+	if spec == nil || status == nil {
+		return fmt.Errorf("passed nil value as spec or status machine network")
+	}
+	if len(*spec) == 0 {
+		// This is a subtlety of JSON marshalling. What happens is that
+		// `var myslice []int` gets marshalled to null, but
+		// `myslice := []int{}` gets marshalled to []
+		// Because of this we want to be explicit that machine networks is an empty list and not null
+		*spec = []configv1.CIDR{}
```

Comment:
The declaration you passed affects only new clusters, but this code also handles CRs that have been created before this field was added. So we are dealing here with 2 scenarios

* CR created now and value is an empty array (because field exists but user did not provide a value)
* CR created long time ago and value is nil (because field did not exist when CR was created)